zilencer/lib/stripe.py

```diff
@@ -178,3 +179,93 @@ def do_subscribe_customer_to_plan(stripe_customer: stripe.Customer, stripe_plan_
                 event_time=timestamp_to_datetime(stripe_subscription.created),
                 requires_billing_update=True,
                 extra_data=ujson.dumps({'quantity': current_seat_count}))
+
+## Process RealmAuditLog
+
+class BillingLogProcessingException(Exception):
+    pass
+
+def is_customer_agnostic_error(error: Exception):
+    # TODO
+    return False
+
+def get_next_log_entry(processor: RealmAuditLogProcessor) -> Optional[RealmAuditLog]:
+    if processor.realm is None:
+        realms_with_processors = RealmAuditLog.objects.values_list('realm', flat=True)
+        query = RealmAuditLog.objects.exclude(realm__in=realms_with_processors)
+    else:
+        query = RealmAuditLog.objects.filter(realm=processor.realm)
+        global_processor = RealmAuditLog.objects.get(realm=None)
+        if global_processor.state == RealmAuditLogProcessor.SKIPPED:
+            query = query.filter(id__lte=global_processor.log_row.id)
+        else:
+            query = query.filter(id__lt=global_processor.log_row.id)
+    return query.filter(id__gt=processor.log_row.id, requires_billing_update=True) \
+                .order_by('id').first()
+
+@catch_stripe_errors
+def do_set_subscription_quantity(
+        customer: Customer, timestamp: int, idempotency_key: str, quantity: int) -> None:
+    stripe_customer = get_stripe_customer(customer.stripe_customer_id)
+    stripe_subscription = extract_current_subscription(stripe_customer)
+    stripe_subscription.quantity = quantity
+    stripe_subscription.proration_date = timestamp
+    stripe_subscription.save(idempotency_key=idempotency_key)
+
+@catch_stripe_errors
+def do_adjust_subscription_quantity(
+        customer: Customer, timestamp: int, idempotency_key: str, delta: int) -> None:
+    stripe_customer = get_stripe_customer(customer.stripe_customer_id)
+    stripe_subscription = extract_current_subscription(stripe_customer)
+    stripe_subscription.quantity = stripe_subscription.quantity + delta
+    stripe_subscription.proration_date = timestamp
+    stripe_subscription.save(idempotency_key=idempotency_key)
+
+def increment_subscription_quantity(
+        customer: Customer, timestamp: int, idempotency_key: str) -> None:
+    return do_adjust_subscription_quantity(customer, timestamp, 1)
+
+def decrement_subscription_quantity(
+        customer: Customer, timestamp: int, idempotency_key: str) -> None:
+    return do_adjust_subscription_quantity(customer, timestamp, -1)
```

Comment:
So, this sort of dead-reckoning of adding +1 or -1 is risky, without some sort of assertion that resulting total matches what it should be.  I'm not sure if such an assertion exists elsewhere in the codebase, but just wanted to flag this as a source of bugs (where an off-by-one bug from one network request failing now is present in every log entry into the future).

Reply:
yup; it's not in this code but is a known follow-up.